### ISSUE TITLE
Complete React 16 support

### ISSUE THREAD

**Issue by ljharb** [airbnb/enzyme#1553]
- [ ] componentDidCatch (#1255)
 - [ ] Portals (#1507, #252)
 - [ ] Fragments (#1213)
 - [ ] Rendering non-jsx (#1149, #1178)
 - [ ] General issues around our use of private APIs (#1512)

**Comment by MarcoNicolodi** [comment]
Isnt #1229 related to this?

**Comment by ljharb** [comment]
@MarcoNicolodi I don't think it is; that issue might either be a bug in v3 or intended behavior, but it's not related to React 16 specifically.

**Comment by petegleeson** [comment]
@ljharb I have verified that #1592 solves the forwardRef problem. 

What is the release plan for these fixes? New versions for each feature or one release when all of these are completed?

Please let me know if there is anything I can do to help out.

**Comment by ljharb** [comment]
I’m hoping to bunch together as many as possible; there’s currently a backlog for me to go through, after which i hope to cut a release.

**Comment by baldwmic** [comment]
@ljharb is there a list of issues you'd like to get merged before releasing a new version on npm? Would love to see this one get released soon: https://github.com/airbnb/enzyme/pull/1513 as my team has already migrated our codebase so just waiting on tests passing.

Perhaps we could help finish up some issues for you to get the release in.

**Comment by simonbuchan** [comment]
What about `createRef()`? I'm hitting that not working in https://github.com/styled-components/styled-components/pull/1720 - even with just pure react + enzyme. Here's the test copied over:

```js
test('enzyme should mount object refs', () => {
  const ref = React.createRef()
  const wrapper = mount(<div ref={ref} />)
  const div = wrapper.find('div').first()
  expect(div).toBeTruthy()
  expect(ref).toHaveProperty('current', div.instance())
})
```

Is this just a dumb mistake I'm making?

**Comment by nemoDreamer** [comment]
Is there any chance of an alpha release, or do _all_ features need to be supported in one go?

**Comment by hvolschenk** [comment]
Not sure if this is bundled together with `<React.Fragment />`, just adding a note in here about `<React.StrictMode />`: https://reactjs.org/docs/strict-mode.html

**Comment by koba04** [comment]
I've created a PR to support `getSnapshotBeforeUpdate` for `shallow` https://github.com/airbnb/enzyme/pull/1657

**Comment by sambokai** [comment]
Can you confirm that the forwardRef problem is solved by facebook/react#12725 ?

**Comment by paska27** [comment]
Also, found inconsistency with `React.Fragment`. Thought to share:

**Describe the bug**
Shallow should find direct child when inside React.Fragment.
(At least it does so for indirect descendants).

**To Reproduce**
```javascript
import * as React from 'react';
import {shallow} from 'enzyme';

const MyCom = () => (
  <div className="container">
    <React.Fragment>
      <span>A span</span>
    </React.Fragment>
  </div>
);

describe('enzyme shallow', () => {
  it('test#1: should find descendant span inside fragment ', () => {
    const myCom = shallow(<MyCom />);
    expect(myCom.find('.container span')).toHaveLength(1);
  });

  // TEST FAILS
  it('test#2: should find direct child span inside fragment', () => {
    const myCom = shallow(<MyCom />);
    expect(myCom.find('.container > span')).toHaveLength(1);
  });
});

```

**Expected behavior**
Expect `test#2` to pass.

**Desktop (please complete the following information):**
 - OS: Ubuntu 16.04 LTS GNOME 
 - Browser: X
 - Version: X

**Additional context**
Tests are failing inside Jest.

**Comment by omril1** [comment]
Is this package dead? Last release to npm was half a year ago, not even patch versions since then.
I would be happy to know what happened.

**Comment by krimple** [comment]
Could it be AirBnB drops support for React Native = no more updates to Enzyme? There was a comment on a post I've referenced below mentioning the trouble they've had when React Native went to React 16 alpha and beta. Can someone comment from the team? https://medium.com/airbnb-engineering/sunsetting-react-native-1868ba28e30a

**Comment by ljharb** [comment]
@krimple no, it absolutely could not be that, and that's ridiculous. All of airbnb.com is in React, and our investment in it has only increased. (the trouble around React 16 and react native is because RN itself unfortunately chose to require an alpha version of React 16 while dropping compat for React 15, and is unrelated to enzyme).

Enzyme currently does not support any of the new features in React 16; that will change in the next release. This project isn't dead in any way; it's just maintained by one person (hi!) who has a personal life and a busy schedule at work.

**Comment by krimple** [comment]
And it turns out the project I was testing used create-react-app and what fixed my issue was upgrading the jest-runner to the latest version. I was digging deep into a rabbit hole and the problem had nothing to do with Enzyme. My apologies again.

**Comment by the-spyke** [comment]
So, how Airbnb managed to survive all those months? Not using any React 16 features? Creating temporary hacks in tests?

**Comment by ljharb** [comment]
@the-spyke not using any React 16 features - we don't use things we can't test.

**Comment by the-spyke** [comment]
@ljharb Which means Airbnb has no need for updated Enzyme. Sounds like chicken-egg problem :-) Thanks for the answer -- now I know what to do.

**Comment by ljharb** [comment]
@the-spyke on the contrary; we have a huge need for it, and are dying to use the new React features. "Need" doesn't directly translate to "available time to work on it", however, which is why it's taken this long. However, now we have two engineers spending actual roadmap time on it, so there will be a release soon.

**Comment by ljharb** [comment]
The following packages are now released:
 - [enzyme v3.4.0](https://www.npmjs.com/package/enzyme)
 - [enzyme-adapter-react-16 v1.2.0](https://www.npmjs.com/package/enzyme-adapter-react-16)
 - [enzyme-adapter-react-16.3 v1.0.0](https://www.npmjs.com/package/enzyme-adapter-react-16.3)
 - [enzyme-adapter-react-16.2 v1.0.0](https://www.npmjs.com/package/enzyme-adapter-react-16.2)
 - [enzyme-adapter-react-16.1 v1.0.0](https://www.npmjs.com/package/enzyme-adapter-react-16.1)
 - [enzyme-adapter-react-15 v1.0.6](https://www.npmjs.com/package/enzyme-adapter-react-15)
 - [enzyme-adapter-react-15.4 v1.0.6](https://www.npmjs.com/package/enzyme-adapter-react-15.4)
 - [enzyme-adapter-react-14 v1.0.6](https://www.npmjs.com/package/enzyme-adapter-react-14)
 - [enzyme-adapter-react-13 v1.0.4](https://www.npmjs.com/package/enzyme-adapter-react-13)
 - [enzyme-adapter-utils v1.5.0](https://www.npmjs.com/package/enzyme-adapter-utils)

**Comment by the-spyke** [comment]
`forwardRef` still doesn't work, by the way.

**Comment by ljharb** [comment]
@simonbuchan fair enough :-) sorry if my reply was harsh.

**Comment by maxkostow** [comment]
Yes it is but there is a bug with github where clicking the check mark unchecks the others so you have to manually edit the markdown.

**Comment by mprobber** [comment]
re: 16.6 `contextType` support - I just put out facebook/react#14329 - it's a PR  for the shallow `react-test-renderer` so that when `contextType` is defined, you can just pass it a context.

**Comment by rodoabad** [comment]
@ljharb I don't see any item for supporting [`React.lazy`](https://reactjs.org/blog/2018/10/23/react-v-16-6.html), I see one for `Suspense` but those two come hand-in-hand.

**Comment by llaryssa** [comment]
Does this issue includes `shallow` rendering support for components that use react hooks?

**Comment by andrewholsted** [comment]
It looks like the Profiler tickbox is still unsupported but the linked issue is closed. Are there plans to add support for this?

**Comment by jimmywongroo** [comment]
I see #14679 has been merged, does that mean the `hooks` box can now be checked?

**Comment by ljharb** [comment]
@jimmywongroo that's the PR on react itself for hooks, so no, that doesn't mean anything.
